runtime/intern.c

```diff
@@ -239,446 +284,395 @@ static void readfloats(struct caml_intern_state* s,
-    for (j = 0; j < STACK_NFIELDS; j++) {
-      s->first_vals[i][j] = Val_unit;
-    }
-  }
-  s->curr_vals = s->first_vals;
-  s->len = INTERN_STACK_INIT_SIZE;
-  s->sp = 0;
-
-  /* Set up GC roots */
-  s->caml__roots_stack.next = CAML_LOCAL_ROOTS;
-  CAML_LOCAL_ROOTS = &s->caml__roots_stack;
-  s->caml__roots_stack.mutexes = 0;
-  s->caml__roots_stack.ntables = 1;
-  s->caml__roots_stack.nitems = INTERN_STACK_INIT_SIZE * STACK_NFIELDS;
-  s->caml__roots_stack.tables[0] = (value*)s->curr_vals;
-}
+  /* Copy item from the old stack to the new stack */
+  memcpy(newstack, s->intern_stack,
+         sizeof(struct intern_item) * sp_offset);
 
-static int stack_is_empty(struct intern_stack* s) {
-  return s->sp == 0;
-}
+  /* Free to old stack if it is not the initial stack */
+  if (s->intern_stack != s->intern_stack_init)
+    caml_stat_free(s->intern_stack);
 
-static void stack_free(struct intern_stack* s) {
-  /* free any memory allocated */
-  if (s->curr_vals != s->first_vals)
-    caml_stat_free(s->curr_vals);
+  s->intern_stack = newstack;
+  s->intern_stack_limit = newstack + newsize;
+  return newstack + sp_offset;
 }
 
-static void stack_realloc(struct intern_stack* s, value save) {
-  CAMLparam1(save);
-  int i;
-  intern_stack_item* new_vals;
-  int new_len = s->len * 2;
-  /* reallocate stack */
-  caml_gc_log("Reallocating intern stack to size %d", new_len);
-  if (new_len >= INTERN_STACK_MAX_SIZE) {
-    caml_gc_log ("Stack overflow in un-marshaling value");
-    stack_free(s);
-    caml_raise_out_of_memory();
-  }
-
-  new_vals = caml_stat_alloc(new_len * STACK_NFIELDS * sizeof(value));
-
-  for (i = 0; i < s->sp; i++) {
-    STACK_VAL(new_vals, i) = STACK_VAL(s->curr_vals, i);
-    STACK_FIELD(new_vals, i) = STACK_FIELD(s->curr_vals, i);
-    STACK_OP(new_vals, i) = STACK_OP(s->curr_vals, i);
-    STACK_ARG(new_vals, i) = STACK_ARG(s->curr_vals, i);
-  }
+/* Convenience macros for requesting operation on the stack */
+#define PushItem(s)                                                     \
+  do {                                                                  \
+    sp++;                                                               \
+    if (sp >= s->intern_stack_limit) sp = intern_resize_stack(s, sp);   \
+  } while(0)
+
+#define ReadItems(s,_dest,_n)                                           \
+  do {                                                                  \
+    if (_n > 0) {                                                       \
+      PushItem(s);                                                      \
+      sp->op = OReadItems;                                              \
+      sp->dest = _dest;                                                 \
+      sp->arg = _n;                                                     \
+    }                                                                   \
+  } while(0)
+
+static void intern_alloc(struct caml_intern_state* s,
```

Comment:
Perhaps `intern_alloc` should be renamed `intern_alloc_workspace` or `intern_alloc_storage` to better distinguish it from the new `intern_alloc_obj` function?